core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c

```diff
@@ -314,41 +283,50 @@ wasi_addr_ip_to_bh_ip_addr_buffer(__wasi_addr_ip_t *addr,
     }
 }
 
+bool
+convert_wasi_clock_id_to_bh_clock_id(__wasi_clockid_t in, bh_clock_id_t *out)
+{
+    switch (in) {
+        case __WASI_CLOCK_MONOTONIC:
+            *out = BH_CLOCK_ID_MONOTONIC;
+            return true;
+        case __WASI_CLOCK_PROCESS_CPUTIME_ID:
+            *out = BH_CLOCK_ID_PROCESS_CPUTIME_ID;
+            return true;
+        case __WASI_CLOCK_REALTIME:
+            *out = BH_CLOCK_ID_REALTIME;
+            return true;
+        case __WASI_CLOCK_THREAD_CPUTIME_ID:
+            *out = BH_CLOCK_ID_THREAD_CPUTIME_ID;
+            return true;
+        default:
+            return false;
+    }
+}
+
 __wasi_errno_t
 wasmtime_ssp_clock_res_get(__wasi_clockid_t clock_id,
                            __wasi_timestamp_t *resolution)
 {
-#ifdef BH_PLATFORM_WINDOWS
-    return __WASI_ENOSYS;
-#else
-    clockid_t nclock_id;
-    if (!convert_clockid(clock_id, &nclock_id))
+    bh_clock_id_t bh_clockid;
+    if (!convert_wasi_clock_id_to_bh_clock_id(clock_id, &bh_clockid))
         return __WASI_EINVAL;
-    struct timespec ts;
-    if (clock_getres(nclock_id, &ts) < 0)
+    if (os_clock_res_get(clock_id, resolution) != BHT_OK)
         return convert_errno(errno);
-    *resolution = convert_timespec(&ts);
-    return 0;
-#endif
+    return __WASI_ESUCCESS;
 }
 
 __wasi_errno_t
 wasmtime_ssp_clock_time_get(__wasi_clockid_t clock_id,
```

Comment:
Sorry for the late comment, did we consider implementing these functions (wasmtime_ssp_clock_time_get and wasmtime_ssp_clock_res_get) directly in each platform? Unlike the filesystem or socket functions, they don't  contain any logic to lookup a fd object from the fd table so it seems they are very small wrapper around the platform implementation anyway.

Also, since `os_clock_time_get` and `os_clock_res_get` are only used in the WASI libc implementation, we could probably define the signature for these functions in terms of WASI types (`__wasi_clockid_t` etc.) to avoid some boilerplate. What do you think?

Reply:
I think that's not been done mainly due to the circular dependency as discussed in  https://github.com/bytecodealliance/wasm-micro-runtime/pull/2585#discussion_r1363526897 

I think we can merge this as it is now, but once your change @zoraaver is merged, we'll make an update to this code and simplify it if possible. How does it sound to you?

Reply:
Agree, I think the PR is in a good state to merge now and we can simplify the code later once #2585  is merged.